src/java.base/share/classes/java/util/zip/GZIPHeaderBuilder.java

```diff
@@ -0,0 +1,361 @@
+     * series of subfields, each of the form:
+     *
+     * +---+---+---+---+==================================+
+     * |SI1|SI2|  LEN  |... LEN bytes of subfield data ...|
+     * +---+---+---+---+==================================+
+     *
+     * SI1 and SI2 provide a subfield ID, typically two ASCII letters
+     * with some mnemonic value
+     *
+     * @param fieldBytes    the data of extra fileds.
+     * @return {@code true} if field data layout is correct.
+     *
+     * @since 18
+     */
+    private boolean verifyExtraFieldLayout(byte[] fieldBytes) {
+        int index = 0;
+        int total = fieldBytes.length;
+        while(index < total) {
+            byte si1 = fieldBytes[index++];
+            byte si2 = fieldBytes[index++];
+            // si1 and si2 should be ASCII
+            if (si1 > 127 || si2 > 127) return false;
+            byte loLen = fieldBytes[index++];
+            byte hiLen = fieldBytes[index++];
+            int len = ((int)hiLen) << 8 | (int)loLen;
+            index += len;
+        }
+        return index == total;
+    }
+
+    /**
+     * This class implements the header of GZIP file which contains members defined
+     * in the RFC 1952 specification
+     *
+     * @since 18
+     *
+     */
+    public record GZIPHeaderData (byte compressMethod, byte flags,
+                                  byte[] extraFieldBytes,
+                                  String filename,
+                                  String fileComment,
+                                  int headerCRC16,
+                                  byte[] headerBytes) {
+        /**
+         * GZIP header magic number.
+         */
+        private static final int GZIP_MAGIC = 0x8b1f;
+
+        /**
+         * Represents the default "unknown" value for OS header, per RFC-1952
+         */
+        private static final byte OS_UNKNOWN = (byte) 255;
+
+        /**
+         * File header flags.
+         *
+         * Per RFC 1952, the header fields contains several members based on flag byte
+         * This flag byte is divided into individual bits as follows:
+         *
+         *                bit 0   FTEXT        file probably ascii text
+         *                bit 1   FHCRC        CRC16 for the gzip header
+         *                bit 2   FEXTRA       extra field present
+         *                bit 3   FNAME        original file name present
+         *                bit 4   FCOMMENT     file comment present
+         *                bit 5   reserved
+         *                bit 6   reserved
+         *                bit 7   reserved
+         *
+         */
+        private static final int FTEXT      = 1;    // Extra text
+        private static final int FHCRC      = 2;    // Header CRC
+        private static final int FEXTRA     = 4;    // Extra field
+        private static final int FNAME      = 8;    // File name
+        private static final int FCOMMENT   = 16;   // File comment
```

Comment:
Since the `flags` record component is public, These probably should be left public as well to benefit users, especially given checking the flags can replace null checks for `extraFieldBytes`, `filename`, and `fileContent`.